x/staking/keeper/delegation.go

```diff
@@ -626,6 +627,152 @@ func (k Keeper) Delegate(
+	transferred := sdk.ZeroDec()
+
+	// sanity checks
+	if !wantShares.IsPositive() {
+		return transferred
+	}
+	validator, found := k.GetValidator(ctx, valAddr)
+	if !found {
+		return transferred
+	}
+	delFrom, found := k.GetDelegation(ctx, fromAddr, valAddr)
+	if !found {
+		return transferred
+	}
+
+	// Check redelegation entry limits while we can still return early.
+	// Assume the worst case that we need to transfer all redelegation entries
+	toRedelegations := k.GetRedelegations(ctx, toAddr, math.MaxUint16)
+	for _, toRedelegation := range toRedelegations {
+		// There's no redelegation index by delegator and dstVal or vice-versa.
+		// The minimum cardinality is to look up by delegator, so scan and skip.
+		if toRedelegation.ValidatorDstAddress != valAddr.String() {
+			continue
+		}
+		fromRedelegation, found := k.GetRedelegation(ctx, fromAddr, sdk.ValAddress(toRedelegation.ValidatorSrcAddress), sdk.ValAddress(toRedelegation.ValidatorDstAddress))
+		if found && len(toRedelegation.Entries)+len(fromRedelegation.Entries) >= int(k.MaxEntries(ctx)) {
+			// avoid types.ErrMaxRedelegationEntries
+			return transferred
+		}
+	}
+
+	// compute shares to transfer, amount left behind
+	transferred = delFrom.Shares
+	if transferred.GT(wantShares) {
+		transferred = wantShares
+	}
+	remaining := delFrom.Shares.Sub(transferred)
+
+	// Update or create the delTo object, calling appropriate hooks
+	delTo, found := k.GetDelegation(ctx, toAddr, validator.GetOperator())
+	if !found {
+		delTo = types.NewDelegation(toAddr, validator.GetOperator(), sdk.ZeroDec())
+	}
+	if found {
+		k.BeforeDelegationSharesModified(ctx, toAddr, validator.GetOperator())
+	} else {
+		k.BeforeDelegationCreated(ctx, toAddr, validator.GetOperator())
+	}
+	delTo.Shares = delTo.Shares.Add(transferred)
+	k.SetDelegation(ctx, delTo)
+	k.AfterDelegationModified(ctx, toAddr, valAddr)
+
+	// Update source delegation
+	if remaining.IsZero() {
+		k.BeforeDelegationRemoved(ctx, fromAddr, valAddr)
+		k.RemoveDelegation(ctx, delFrom)
+	} else {
+		k.BeforeDelegationSharesModified(ctx, fromAddr, valAddr)
+		delFrom.Shares = remaining
+		k.SetDelegation(ctx, delFrom)
+		k.AfterDelegationModified(ctx, fromAddr, valAddr)
+	}
+
+	// If there are not enough remaining shares to be responsible for
+	// the redelegations, transfer some redelegations.
+	// For instance, if the original delegation of 300 shares to validator A
+	// had redelegations for 100 shares each from validators B, C, and D,
+	// and if we're transferring 175 shares, then we might keep the redelegation
+	// from B, transfer the one from D, and split the redelegation from C
+	// keeping a liability for 25 shares and transferring one for 75 shares.
+	// Of course, the redelegations themselves can have multiple entries for
+	// different timestamps, so we're actually working at a finer granularity.
+	redelegations := k.GetRedelegations(ctx, fromAddr, math.MaxUint16)
```

Comment:
ditto use iterator

Reply:
It looks like you break the contract if you modify the section of the store you're iterating over.  See comment on `Iterator()` method of `KVStore` in `store/types/store.go`.